### PR TITLE
RFC - Full Wallet Lifecycle

### PR DESCRIPTION
**Pull request by yeastplume**
Replaces https://github.com/mimblewimble/grin-rfcs/pull/11 (for git reasons, conversation in previous issue is still valid)

Adds detailed description of potential security model and sequence diagrams demonstrating.

[Link to rendered doc](https://github.com/yeastplume/grin-rfcs/blob/wallet_lifecycle/text/0000-full-wallet-lifecycle.md)

### CODE REVIEW

**Comment by yeastplume** [comment]
[Rendered link to RFC document](https://github.com/yeastplume/grin-rfcs/blob/wallet_lifecycle/text/0000-full-wallet-lifecycle.md)

**Comment by yeastplume** [comment]
In accordance with our RFC process, as of today this can be considered in Final Comment Period with a disposition of merge

File: text/0000-full-wallet-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+creates a seed file, stores the resulting data files in the directory specified in `grin-wallet.toml` (`~/.grin/main/wallet_data` by default)
+and initialises the lmdb database.
+
+It should be possible to run `grin-wallet owner_api` or invoke the API directly from a linked binary without having instantiated a wallet.
+
+## Security Model
+
+Given that the Wallet's Owner API needs to deal with sensitive data such as passwords and seed phrases, the API will be enhanced with a new ECDH and Token-based security model, the primary goals of which are to:
+
+* Ensure sensitive data such as passwords or seed phrases are always end-to-end encrypted between the client and Owner API server, regardless of what higher-level protocols are used during the exchange.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: Missing "the" in front of Owner API server?

File: text/0000-full-wallet-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+* Minimize the potential for damage that can be done by a third party listening on the exchange between a wallet client and its corresponding server.
+* Ensure that sensitive data such as passwords or seed phrases are not resident in server-side memory any longer than they absolutely need to be.
+
+Note that this mode of operation is primarly intended for use over the JSON-RPC API, which supports many different architectural possiblities. Clients that link libraries directly and keep all sensitive data in the same process would see less benefit from this scheme, and an alternative model which doesn't encrypt any sensitive data is provided. Further, authors of existing wallets will need time to consider and/or implement the added complexity needed on the client-side to support ECDH and encryption. It's therefore proposed that the Owner API initially provide the new "SecureAPI" mode as an optional feature, with wallet authors strongly encouraged to make use of it. Support for the "InsecureAPI" model can be maintained indefinitely for directly-linked wallets, and for the JSON-RPC API until a cut-off release at some point in the future.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: > Note that this mode of operation is primarly intended

Not English native but isn't it primarily?

Also

> architectural possiblities

architectural possibilities.






Missing "s" at:
> It's therefore proposed that the Owner API initially provide

File: text/0000-full-wallet-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+ECDH will use secp256k1 for key agreement, while encryption of JSON-RPC requests and responses will be performed via the AES256 Encryption standard.
+
+#### Security Mode Selection
+
+To initialize SecureAPI Mode, clients will generate an EC keypair using the secp256k1 curve, and provide the public key to the Owner API server via a new `init_api_secure` method. Both client and server will calculate the shared key, and store this key for the remainder of the session. The sequence of operations is outlined below:
+
+![init_api_secure](../assets/0000/0000-dh-init.svg)
+
+Once `init_api_secure` is called the API will assume that subsequent API requests will be encrypted with the shared secret, and that that all subsequent API calls (other than `open_wallet`) will be accompanied with a valid encrypted token derived during the call to the `open_wallet` function. This assumption will remain until the server process exits or a call to a corresponding `close_api_secure` function is called.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: Missing comma after "Once `init_api_secure` is called". Not sure about the need for a comma before the ", and that that all".

File: text/0000-full-wallet-lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+The shared secret can be refreshed by the client at any time with another call to `init_api_secure`. Closing a wallet via the `close_wallet` function does not regenerate the shared secret but does invalidate the token and drops the XORed seed from memory.
+
+##### Legacy support
+
+If `init_api_secure` is not called, the mode of operation for the JSON-RPC API will be assumed to work as currently, i.e. requests and responses are unencrypted, the wallet stores its full seed in-memory between requests and the providing of a Token with each request is not requred. However, the new lifecycle functions described in this RFC, which deal with highly sensitive data such as passwords and master keys, will not be available via the JSON-RPC API unless `init_api_secure` has been called. This setup should allow existing wallets to continue working as-is until a cutoff release for legacy mode is determined.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: >providing of a Token with each request is not requred

providing of a token with each request is not required.

File: text/0000-full-wallet-lifecycle.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+#### Opening a Wallet in SecureAPI Mode
+
+Opening a wallet in SecureAPI mode consists of encrypting a request to `open_wallet` (which contains the wallet password) with the shared secret `s`. The request is decrypted in the JSON-RPC layer, and the password is used in the wallet backend to unlock the wallet master seed. The master seed is stored XORed against a randomly-generated token T, which is returned to the client in an encrypted response for inclusion in all further API calls. T is valid for the lifetime of the process, or until a corresponding call to `close_wallet`.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: > The request is decrypted in the JSON-RPC layer, and the password

Unnecessary comma?

> process, or until a corresponding call to `close_wallet`.

Same comment.

File: assets/0000/0000-open-wallet.svg [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+title Opening a Wallet in Secure Mode via JSON-RPC API
+
+== Open Wallet ==
+client -> client: Supply password field **pw** in **req** as **req.pw**
+client -> client: enc(**req**, **s**) -> **reqEnc**
+client -> jsonrpc: open_wallet(**reqEnc**)
+jsonrpc -> jsonrpc: dec(**reqEnc**, **s**) -> **req**
+jsonrpc -> rustapi: open_wallet(**req**)
+rustapi -> rustapi: Generate Random Token (**t**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet.svg]: > Generate Random Token

In lowercase to be consistent with the rest of the diagram?

File: assets/0000/0000-open-wallet.svg [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+rustapi -> backend: Open Wallet(**t**, **req.pw**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet.svg]: Same comment as above.

File: assets/0000/0000-open-wallet.puml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+title Opening a Wallet in Secure Mode via JSON-RPC API
+
+== Open Wallet ==
+client -> client: Supply password field **pw** in **req** as **req.pw**
+client -> client: enc(**req**, **s**) -> **reqEnc**
+client -> jsonrpc: open_wallet(**reqEnc**)
+jsonrpc -> jsonrpc: dec(**reqEnc**, **s**) -> **req**
+jsonrpc -> rustapi: open_wallet(**req**)
+rustapi -> rustapi: Generate Random Token (**t**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet.puml]: > Generate Random Token

In lowercase to be consistent with the rest of the diagram?

File: assets/0000/0000-open-wallet.puml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+rustapi -> backend: Open Wallet(**t**, **req.pw**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet.puml]: Same as above.

File: assets/0000/0000-api-call-direct.puml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+participant "Wallet Backend" as backend
+
+title Wallet API Call (Directly Linked)
+
+== API Call ==
+client -> client: Stored token **t**
+client -> rustapi: call_api_function(**t**, **other_params**)
+rustapi -> backend: retrieve_seed(**t**)
+backend -> backend: calc(**t** ^ **seedT**) -> **seed**
+rustapi -> backend: Internal Wallet Calls
```
Comment (quentinlesceller) [inline assets/0000/0000-api-call-direct.puml]: > Internal Wallet Calls

Lowercase to be consistent with the rest of the content?

File: assets/0000/0000-api-call-direct.svg [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+participant "Wallet Backend" as backend
+
+title Wallet API Call (Directly Linked)
+
+== API Call ==
+client -> client: Stored token **t**
+client -> rustapi: call_api_function(**t**, **other_params**)
+rustapi -> backend: retrieve_seed(**t**)
+backend -> backend: calc(**t** ^ **seedT**) -> **seed**
+rustapi -> backend: Internal Wallet Calls
```
Comment (quentinlesceller) [inline assets/0000/0000-api-call-direct.svg]: Lowercase to be consistent with the rest of the content?

File: assets/0000/0000-api-call.puml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+== API Call ==
+client -> client: Supply stored token **t** in **req** as **req.t**
+client -> client: enc(**req**, **s**) -> **reqEnc**
+client -> jsonrpc: call_api_function(**reqEnc**)
+jsonrpc -> jsonrpc: dec(**reqEnc**, **s**) -> **req**
+jsonrpc -> rustapi: call_api_function(**req**)
+rustapi -> backend: retrieve_seed(**req.t** as **t**)
+backend -> backend: calc(**t** ^ **seedT**) -> **seed**
+rustapi -> backend: Internal Wallet Calls
```
Comment (quentinlesceller) [inline assets/0000/0000-api-call.puml]: Lowercase to be consistent with the rest of the content?

File: assets/0000/0000-api-call.svg [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+== API Call ==
+client -> client: Supply stored token **t** in **req** as **req.t**
+client -> client: enc(**req**, **s**) -> **reqEnc**
+client -> jsonrpc: call_api_function(**reqEnc**)
+jsonrpc -> jsonrpc: dec(**reqEnc**, **s**) -> **req**
+jsonrpc -> rustapi: call_api_function(**req**)
+rustapi -> backend: retrieve_seed(**req.t** as **t**)
+backend -> backend: calc(**t** ^ **seedT**) -> **seed**
+rustapi -> backend: Internal Wallet Calls
```
Comment (quentinlesceller) [inline assets/0000/0000-api-call.svg]: Lowercase to be consistent with the rest of the content?

File: assets/0000/0000-dh-init.puml [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+participant "JSON-RPC API" as jsonrpc
+participant "Internal Rust API" as rustapi
+participant "Wallet Backend" as backend
+
+title Wallet JSON-RPC API ECDH Initialization
+
+== Initialization (ECDH/secp256k1)==
+client -> client: Calc Pub ECDH Key (**cP**)
+client -> jsonrpc: init_api_secure (**cP**)
+jsonrpc -> jsonrpc: Calc Pub ECDH Key (**sP**)
```
Comment (quentinlesceller) [inline assets/0000/0000-dh-init.puml]: Same comment as above for the two lines.

File: assets/0000/0000-dh-init.svg [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+participant "JSON-RPC API" as jsonrpc
+participant "Internal Rust API" as rustapi
+participant "Wallet Backend" as backend
+
+title Wallet JSON-RPC API ECDH Initialization
+
+== Initialization (ECDH/secp256k1)==
+client -> client: Calc Pub ECDH Key (**cP**)
+client -> jsonrpc: init_api_secure (**cP**)
+jsonrpc -> jsonrpc: Calc Pub ECDH Key (**sP**)
```
Comment (quentinlesceller) [inline assets/0000/0000-dh-init.svg]: Same as above.

File: assets/0000/0000-open-wallet-direct.puml [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+actor "Client" as client
+participant "Internal Rust API" as rustapi
+participant "Wallet Backend" as backend
+
+title Opening a Wallet (Directly Linked)
+
+== Open Wallet ==
+client -> client: Supply password **pw**
+client -> rustapi: open_wallet(**pw**, **other_params**)
+rustapi -> rustapi: Generate Random Token (**t**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet-direct.puml]: Lowercase to be consistent with the rest of the content?

File: assets/0000/0000-open-wallet-direct.puml [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+rustapi -> backend: open_Wallet(**t**, **pw**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet-direct.puml]: typo `open_wallet`.

File: assets/0000/0000-open-wallet-direct.svg [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+actor "Client" as client
+participant "Internal Rust API" as rustapi
+participant "Wallet Backend" as backend
+
+title Opening a Wallet (Directly Linked)
+
+== Open Wallet ==
+client -> client: Supply password **pw**
+client -> rustapi: open_wallet(**pw**, **other_params**)
+rustapi -> rustapi: Generate Random Token (**t**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet-direct.svg]: Same as above.

File: assets/0000/0000-open-wallet-direct.svg [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+rustapi -> backend: open_Wallet(**t**, **pw**)
```
Comment (quentinlesceller) [inline assets/0000/0000-open-wallet-direct.svg]: `open_wallet`.

File: text/0000-full-wallet-lifecycle.md [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+
+#### Calling API functions in SecureAPI Mode
+
+Calls to each API function proceed as per a call to `open_wallet`, however each encrypted request must contain the token provided by the `open_wallet` call. The token is XORed against the stored XORed seed to recover the original seed by the backend for the duration of each call, and the seed value is dropped and zeroed from memory when each call returns.
+
+![api_call](../assets/0000/0000-api-call.svg)
+
+### Directly Linked Wallets
+
+Wallets that link the wallet API directly will not be required to encrypt parameters, as there would be little benefit to doing so within a single process. However, for consistency they will be expected to store and supply a token to each API call. The modified workflow for a linked wallet is outlined below:
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: > However, for consistency they...

However, for consistency, they...

File: text/0000-full-wallet-lifecycle.md [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+
+![open-wallet-direct](../assets/0000/0000-open-wallet-direct.svg)
+![api-call-direct](../assets/0000/0000-api-call-direct.svg)
+
+'Legacy' support will not be provided for directly-linked wallets on release of the features described in this RFC. Is is expected wallet authors will need to update their code to store and supply the token with each request.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: > Is is expected wallet authors

It is expected wallet authors

File: text/0000-full-wallet-lifecycle.md [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+
+## New Lifecycle API Functions
+
+The functions as shown here are for illustrative purposes, and their signatures will change during implementation.
+
+* `OwnerAPI::init_api_secure(pubkey: Secp256k1Point) -> Result<pubkey: Secp256K1Point, Error>
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: Missing ` at the end of the line.

File: text/0000-full-wallet-lifecycle.md [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+
+### API only
+
+Note that this RFC does not propose making user-facing changes to the existing CLI wallet to invoke these functions. It's expected that the existing cli functionality will be modified to invoke the new API functions.
+
+### Implementation notes
+
+Although this document doesn't attempt to outline implementation, a few notes to consider for the implementor:
+
+* Currently, the code that deals with wallet initialization and seed management sits outside the wallet APIs, in the `impls` crate, (denoting they're implementation specific). The implementation should attempt to refactor traits from these hard implementations into a new interface, similar to the existing WalletBackend and NodeClient interfaces (WalletLifecycleManager, for instance). The implementation within `impls` will then become an implementation of that trait, and can be substituted by wallet authors with their own implementations.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: > (WalletLifecycleManager, for instance)

Why comma here?

> trait, and can 

Same

File: text/0000-full-wallet-lifecycle.md [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+    - Outputs a `grin-wallet.toml` file into current top-level system wallet directory
+    - Optionally takes wallet configuration structure to override defaults in the grin-wallet.toml file
+* `OwnerAPI::create_wallet(name: Option<String>, mnemonic: Option<ZeroingString>, mnemonic_length: usize, password: ZeroingString) -> Result<(), libwallet::Error>`
+    - Creates and initializes a new wallet
+    - Initializes seed from given mnemonic if given, random seed otherwise
+    - Should error appropriately if the wallet already exists
+    - The 'name' parameter is included for future use as in `open_wallet` above.
+* `OwnerAPI::open_wallet(name: Option<String>, password: String) -> Result<t:Token, libwallet::Error>`
+    - Opens a wallet and sets it as the 'active' wallet. All further API commands will be performed against this wallet.
+    - 'Opens' the wallet seed in memory, stored XORd against a new token. The token is to be returned to the client for use in all further API calls.
```
Comment (quentinlesceller) [inline text/0000-full-wallet-lifecycle.md]: > XORd 

XORed.

**Comment by yeastplume** [comment]
Thanks! Comments noted and doc updated.

**Pull request merged**